api/v1/_email/PKPEmailHandler.inc.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+/**
+ * @file api/v1/_email/PKPEmailHandler.inc.php
+ *
+ * Copyright (c) 2014-2020 Simon Fraser University
+ * Copyright (c) 2003-2020 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class PKPEmailHandler
+ * @ingroup api_v1_announcement
+ *
+ * @brief Handle API requests for announcement operations.
+ *
+ */
+use \Illuminate\Queue\Capsule\Manager as Queue;
+use Illuminate\Database\Capsule\Manager as Capsule;
+use \Psr\Http\Message\ServerRequestInterface;
+
+import('lib.pkp.classes.handler.APIHandler');
+import('classes.core.Services');
+
+class PKPEmailHandler extends APIHandler {
+
+	/** Number of emails to send in each job */
+	const EMAILS_PER_JOB = 100;
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		$this->_handlerPath = '_email';
+		$this->_endpoints = [
+			'POST' => [
+				[
+					'pattern' => $this->getEndpointPattern(),
+					'handler' => [$this, 'create'],
+					'roles' => [ROLE_ID_SITE_ADMIN, ROLE_ID_MANAGER],
+				],
+			],
+			'PUT' => [
+				[
+					'pattern' => $this->getEndpointPattern() . '/{queueId}',
+					'handler' => [$this, 'process'],
+					'roles' => [ROLE_ID_SITE_ADMIN, ROLE_ID_MANAGER],
+				],
+			],
+		];
+		parent::__construct();
+	}
+
+	/**
+	 * @copydoc PKPHandler::authorize
+	 */
+	public function authorize($request, &$args, $roleAssignments) {
+		import('lib.pkp.classes.security.authorization.PolicySet');
+		$rolePolicy = new PolicySet(COMBINING_PERMIT_OVERRIDES);
+
+		import('lib.pkp.classes.security.authorization.RoleBasedHandlerOperationPolicy');
+		foreach ($roleAssignments as $role => $operations) {
+			$rolePolicy->addPolicy(new RoleBasedHandlerOperationPolicy($request, $role, $operations));
+		}
+		$this->addPolicy($rolePolicy);
+
+		return parent::authorize($request, $args, $roleAssignments);
+	}
+
+	/**
+	 * Create a jobs queue to send a bulk email to users in one or
+	 * more user groups
+	 *
+	 * @param ServerRequestInterface $slimRequest
+	 * @param APIResponse $response
+	 * @param array $args arguments
+	 * @return APIResponse
+	 */
+	public function create(ServerRequestInterface $slimRequest, APIResponse $response, array $args) {
+		$contextId = $this->getRequest()->getContext()->getId();
+
+		$requestParams = $slimRequest->getParsedBody();
+
+		$params = [];
+		foreach ($requestParams as $param => $val) {
+			switch ($param) {
+				case 'userGroupIds':
+					if (!is_array($val)) {
```

Comment:
Off-topic bikeshedding, but how do you feel about using PHP's `urlParameter[]` syntax to specify multiple values? Then you can just use a standard `$vals = (array) $request->getUserVar('urlParameter')` to get the list as an array, no if-it's-not-an-array-then-do-this fiddling required.

Reply:
I'd definitely like to find a better way to do this but in this case I was running into an annoying bug because the form may submit with an empty string as a value, and this was being converted into `[0]` by the `intval`. I just bashed my way through it for now. :)

Reply:
:+1: :+1: